File: src/count_min_sketch.c
/*******************************************************************************
***     Author: Tyler Barrus
***     email:  barrust@gmail.com
***     Version: 0.1.5
***     License: MIT 2017
*******************************************************************************/

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <limits.h>
#include <inttypes.h>       /* PRIu64 */
#include <math.h>
#include "count_min_sketch.h"

#define LOG_TWO 0.6931471805599453

/* private functions */
static int __setup_cms(CountMinSketch *cms, uint32_t width, uint32_t depth, double error_rate, double confidence, cms_hash_function hash_function);
static void __write_to_file(CountMinSketch *cms, FILE *fp, short on_disk);
static void __read_from_file(CountMinSketch *cms, FILE *fp, short on_disk, char *filename);
static uint64_t* __default_hash(int32_t num_hashes, char *key);
static uint64_t __fnv_1a(char *key);
int __compare (const void * a, const void * b);
int32_t __safe_add(int32_t a, int32_t b);
int32_t __safe_sub(int32_t a, int32_t b);


int cms_init_optimal_alt(CountMinSketch *cms, double error_rate, double confidence, cms_hash_function hash_function) {
    /* https://cs.stackexchange.com/q/44803 */
    uint32_t width = ceil(2 / error_rate);
    uint32_t depth = ceil((-1 * log(1 - confidence)) / LOG_TWO);
    return __setup_cms(cms, width, depth, error_rate, confidence, hash_function);
}

int cms_init_alt(CountMinSketch *cms, uint32_t width, uint32_t depth, cms_hash_function hash_function) {
    double confidence = 1 - (1 / pow(2, depth));
    double error_rate = 2 / (double) width;
    return __setup_cms(cms, width, depth, error_rate, confidence, hash_function);
}

int cms_destroy(CountMinSketch *cms) {
    free(cms->bins);
    cms->width = 0;
    cms->depth = 0;
    cms->confidence = 0.0;
    cms->error_rate = 0.0;
    cms->elements_added = 0;
    cms->hash_function = NULL;
    cms->bins = NULL;

    return CMS_SUCCESS;
}

int cms_clear(CountMinSketch *cms) {
    uint32_t i, j = cms->width * cms->depth;
    for (i = 0; i < j; i++) {
        cms->bins[i] = 0;
    }
    cms->elements_added = 0;
    return CMS_SUCCESS;
}

int32_t cms_add_inc_alt(CountMinSketch *cms, uint64_t* hashes, int32_t num_hashes, uint32_t x) {
    if (num_hashes < cms->depth) {
        fprintf(stderr, "Insufficient hashes to complete the addition of the element to the count-min sketch!");
        return CMS_ERROR;
    }
    int i, num_add = INT_MAX;
    for (i = 0; i < cms->depth; i++) {
        uint64_t bin = (hashes[i] % cms->width) + (i * cms->width);
        cms->bins[bin] = __safe_add(cms->bins[bin], x);
        /* currently a standard min strategy */
        if (cms->bins[bin] < num_add) {
            num_add = cms->bins[bin];
        }
    }
    cms->elements_added = __safe_add(cms->elements_added, (int64_t)x);
    return num_add;
}

int32_t cms_add_inc(CountMinSketch *cms, char* key, unsigned int x) {
    uint64_t* hashes = cms_get_hashes(cms, key);
    int32_t num_add = cms_add_inc_alt(cms, hashes, cms->depth, x);
    free(hashes);
    return num_add;
}

int32_t cms_remove_inc_alt(CountMinSketch *cms, uint64_t* hashes, int num_hashes, unsigned int x) {
    if (num_hashes < cms->depth) {
        fprintf(stderr, "Insufficient hashes to complete the removal of the element to the count-min sketch!");
        return CMS_ERROR;
    }
    int32_t i, num_add = INT_MAX;
    for (i = 0; i < cms->depth; i++) {
        uint32_t bin = (hashes[i] % cms->width) + (i * cms->width);
        cms->bins[bin] = __safe_sub(cms->bins[bin], x);
        if (cms->bins[bin] < num_add) {
            num_add = cms->bins[bin];
        }
    }
    cms->elements_added = __safe_sub(cms->elements_added, (int64_t)x);
    return num_add;
}

int32_t cms_remove_inc(CountMinSketch *cms, char* key, uint32_t x) {
    uint64_t* hashes = cms_get_hashes(cms, key);
    int32_t num_add = cms_remove_inc_alt(cms, hashes, cms->depth, x);
    free(hashes);
    return num_add;
}

int32_t cms_check_alt(CountMinSketch *cms, uint64_t* hashes, int num_hashes) {
    if (num_hashes < cms->depth) {
        fprintf(stderr, "Insufficient hashes to complete the min lookup of the element to the count-min sketch!");
        return CMS_ERROR;
    }
    int32_t i, num_add = INT_MAX;
    for (i = 0; i < cms->depth; i++) {
        uint32_t bin = (hashes[i] % cms->width) + (i * cms->width);
        if (cms->bins[bin] < num_add) {
            num_add = cms->bins[bin];
        }
    }
    return num_add;
}

int32_t cms_check(CountMinSketch *cms, char* key) {
    uint64_t* hashes = cms_get_hashes(cms, key);
    int32_t num_add = cms_check_alt(cms, hashes, cms->depth);
    free(hashes);
    return num_add;
}

int32_t cms_check_mean_alt(CountMinSketch *cms, uint64_t* hashes, int num_hashes) {
    if (num_hashes < cms->depth) {
        fprintf(stderr, "Insufficient hashes to complete the mean lookup of the element to the count-min sketch!");
        return CMS_ERROR;
    }
    int32_t i, num_add = 0;
    for (i = 0; i < cms->depth; i++) {
        uint32_t bin = (hashes[i] % cms->width) + (i * cms->width);
        num_add += cms->bins[bin];
    }
    return num_add / cms->depth;
}

int32_t cms_check_mean(CountMinSketch *cms, char* key) {
    uint64_t* hashes = cms_get_hashes(cms, key);
    int32_t num_add = cms_check_mean_alt(cms, hashes, cms->depth);
    free(hashes);
    return num_add;
}

int32_t cms_check_mean_min_alt(CountMinSketch *cms, uint64_t* hashes, int num_hashes) {
    if (num_hashes < cms->depth) {
        fprintf(stderr, "Insufficient hashes to complete the mean-min lookup of the element to the count-min sketch!");
        return CMS_ERROR;
    }
    int32_t i, num_add = 0;
    int64_t* mean_min_values = calloc(cms->depth, sizeof(long));
    for (i = 0; i < cms->depth; i++) {
        uint32_t bin = (hashes[i] % cms->width) + (i * cms->width);
        int32_t val = cms->bins[bin];
        mean_min_values[i] = val - ((cms->elements_added - val) / (cms->width - 1));
    }
    // return the median of the mean_min_value array... need to sort first
    qsort(mean_min_values, cms->depth, sizeof(long), __compare);
    int32_t n = cms->depth;
    if (n % 2 == 0) {
        num_add = (mean_min_values[n/2] + mean_min_values[n/2 - 1]) / 2;
    } else {
        num_add = mean_min_values[n/2];
    }
    free(mean_min_values);
    return num_add;
}

int32_t cms_check_mean_min(CountMinSketch *cms, char* key) {
    uint64_t* hashes = cms_get_hashes(cms, key);
    int32_t num_add = cms_check_mean_min_alt(cms, hashes, cms->depth);
    free(hashes);
    return num_add;
}

uint64_t* cms_get_hashes_alt(CountMinSketch *cms, int num_hashes, char* key) {
    return cms->hash_function(num_hashes, key);
}

int cms_export(CountMinSketch *cms, char* filepath) {
    FILE *fp;
    fp = fopen(filepath, "w+b");
    if (fp == NULL) {
        fprintf(stderr, "Can't open file %s!\n", filepath);
        return CMS_ERROR;
    }
    __write_to_file(cms, fp, 0);
    fclose(fp);
    return CMS_SUCCESS;
}

int cms_import_alt(CountMinSketch *cms, char* filepath, cms_hash_function hash_function) {
    FILE *fp;
    fp = fopen(filepath, "r+b");
    if (fp == NULL) {
        fprintf(stderr, "Can't open file %s!\n", filepath);
        return CMS_ERROR;
    }
    __read_from_file(cms, fp, 0, NULL);
    cms->hash_function = (hash_function == NULL) ? __default_hash : hash_function;
    fclose(fp);
    return CMS_SUCCESS;
}

/*******************************************************************************
*    PRIVATE FUNCTIONS
*******************************************************************************/
static int __setup_cms(CountMinSketch *cms, unsigned int width, unsigned int depth, double error_rate, double confidence, cms_hash_function hash_function) {
    cms->width = width;
    cms->depth = depth;
    cms->confidence = confidence;
    cms->error_rate = error_rate;
    cms->elements_added = 0;
    cms->bins = calloc(width * depth, sizeof(int32_t));
    cms->hash_function = (hash_function == NULL) ? __default_hash : hash_function;

    return CMS_SUCCESS;
}

static void __write_to_file(CountMinSketch *cms, FILE *fp, short on_disk) {
    unsigned long long i, length = cms->depth * cms->width;
    if (on_disk == 0) {
        for (i = 0; i < length; i++) {
            fwrite(&cms->bins[i], sizeof(int32_t), 1, fp);
        }
    } else {
        // TODO: decide if this should be done directly on disk or not
        // will need to write out everything by hand
        // uint64_t i;
        // int q = 0;
        // for (i = 0; i < length; i++) {
        //     fwrite(&q, sizeof(int), 1, fp);
        // }
    }
    fwrite(&cms->width, sizeof(int32_t), 1, fp);
    fwrite(&cms->depth, sizeof(int32_t), 1, fp);
    fwrite(&cms->elements_added, sizeof(int64_t), 1, fp);
}

static void __read_from_file(CountMinSketch *cms, FILE *fp, short on_disk, char *filename) {
    /* read in the values from the file before getting the sketch itself */
    int offset = (sizeof(int32_t) * 2) + sizeof(long);
    fseek(fp, offset * -1, SEEK_END);
    size_t read;
    read = fread(&cms->width, sizeof(int32_t), 1, fp);
    read = fread(&cms->depth, sizeof(int32_t), 1, fp);
    cms->confidence = 1 - (1 / pow(2, cms->depth));
    cms->error_rate = 2 / (double) cms->width;
    read = fread(&cms->elements_added, sizeof(int64_t), 1, fp);

    rewind(fp);
    long length = cms->width * cms->depth;
    if (on_disk == 0) {
        cms->bins = malloc(length * sizeof(int32_t));
        read = fread(cms->bins, sizeof(int32_t), length, fp);
        if (read != length) {
            perror("__read_from_file: ");
            exit(1);
        }
    } else {
        // TODO: decide if this should be done directly on disk or not
    }
}

static uint64_t* __default_hash(int num_hashes, char *key) {
    uint64_t *results = calloc(num_hashes, sizeof(uint64_t));
    int i;
    char *str = calloc(17, sizeof(char));  // largest value is 7FFF,FFFF,FFFF,FFFF
    for (i = 0; i < num_hashes; i++) {
        if (i == 0) {
            results[i] = __fnv_1a(key);
        } else {
            uint64_t prev = results[i-1];
            memset(str, 0, 17);
            sprintf(str, "%" PRIx64 "", prev);
            results[i] = __fnv_1a(str);
        }
    }
    free(str);
    return results;
}

static uint64_t __fnv_1a(char *key) {
    // FNV-1a hash (http://www.isthe.com/chongo/tech/comp/fnv/)
    int i, len = strlen(key);
    uint64_t h = 14695981039346656073ULL; // FNV_OFFSET 64 bit
    for (i = 0; i < len; i++){
            h = h ^ (unsigned char) key[i];
            h = h * 1099511628211ULL; // FNV_PRIME 64 bit
    }
    return h;
}


int __compare (const void * a, const void * b) {
  return ( *(long*)a - *(long*)b );
}


int32_t __safe_add(int a, int b) {
    /* use the gcc macro if compiling with GCC, otherwise, simple overflow check */
    int32_t c = 0;
    #ifdef __GNUC__
        int bl = __builtin_add_overflow(a, b, &c);
        if (bl != 0) {
            c = INT_MAX;
        }
    #else
        if (b < INT_MIN + a) {
            c = INT_MIN;
        } else {
            c = a - b;
        }
    #endif

    return c;
}

int32_t __safe_sub(int32_t a, int32_t b) {
    /* use the gcc macro if compiling with GCC, otherwise, simple overflow check */
    int32_t c = 0;
    #ifdef __GNUC__
        int32_t bl = __builtin_sub_overflow(a, b, &c);
        if (bl != 0) {
            c = INT_MAX;
        }
    #else
        if (b > INT_MAX - a) {
            c = INT_MAX;
        } else {
            c = a + b;
        }
    #endif

    return c;
}

File: python/countminsketch.py
''' Count-Min Sketch python implementation
    License: MIT
    Author: Tyler Barrus (barrust@gmail.com)
    URL: https://github.com/barrust/count-min-sketch
'''
from __future__ import (print_function)
import os
from struct import (pack, unpack, calcsize)
import math


class CountMinSketch(object):
    ''' Count-Min Sketch class '''

    def __init__(self, width=None, depth=None, confidence=None,
                 error_rate=None, filepath=None, hash_function=None):
        ''' default initilization function '''
        # default values
        self.__width = 0
        self.__depth = 0
        self.__confidence = 0.0
        self.__error_rate = 0.0
        self.__elements_added = 0
        # for python2 and python3 support
        self.__int32_t_min = -2147483648
        self.__int32_t_max = 2147483647
        self.__int64_t_min = -9223372036854775808
        self.__int64_t_max = 9223372036854775807
        self.__uint64_t_max = 2 ** 64

        if width is not None and depth is not None:
            self.__width = width
            self.__depth = depth
            self.__confidence = 1 - (1 / math.pow(2, depth))
            self.__error_rate = 2 / width
            self._bins = [0] * (self.__width * self.__depth)
        elif confidence is not None and error_rate is not None:
            self.__confidence = confidence
            self.__error_rate = error_rate
            self.__width = math.ceil(2 / error_rate)
            numerator = (-1 * math.log(1 - confidence))
            self.__depth = math.ceil(numerator / 0.6931471805599453)
            self._bins = [0] * (self.__width * self.__depth)
        elif filepath is not None:
            self.load(filepath, hash_function)

        if hash_function is None:
            self._hash_function = self.__default_hash
        else:
            self._hash_function = hash_function

    @property
    def width(self):
        ''' get width '''
        return self.__width

    @property
    def depth(self):
        ''' get depth '''
        return self.__depth

    @property
    def confidence(self):
        ''' get confidence '''
        return self.__confidence

    @property
    def error_rate(self):
        ''' get error rate '''
        return self.__error_rate

    @property
    def elements_added(self):
        ''' get elements added '''
        return self.__elements_added

    def clear(self):
        ''' reset the count-min sketch to empty '''
        self.__elements_added = 0
        for i, _ in enumerate(self._bins):
            self._bins[i] = 0

    def hashes(self, key, depth=None):
        ''' return the hashes for the passed in key '''
        t_depth = self.__depth if depth is None else depth
        return self._hash_function(key, t_depth)

    def add(self, key, num_els=1):
        ''' add element 'key' to the count-min sketch 'x' times '''
        hashes = self.hashes(key)
        return self.add_alt(hashes, num_els)

    def add_alt(self, hashes, num_els=1):
        ''' add an element by using the hashes '''
        res = self.__int32_t_max
        for i, val in enumerate(hashes):
            t_bin = (val % self.__width) + (i * self.__width)
            self._bins[t_bin] += num_els
            if self._bins[t_bin] > self.__int32_t_max:
                self._bins[t_bin] = self.__int32_t_max
            if self._bins[t_bin] < res:
                res = self._bins[t_bin]

        self.__elements_added += num_els
        if self.__elements_added > self.__int32_t_max:
            self.__elements_added = self.__int32_t_max

        return res

    def remove(self, key, num_els=1):
        ''' remove element 'key' from the count-min sketch 'x' times '''
        hashes = self.hashes(key)
        return self.remove_alt(hashes, num_els)

    def remove_alt(self, hashes, num_els=1):
        ''' remove an element by using the hashes '''
        res = self.__int32_t_max
        for i, val in enumerate(hashes):
            t_bin = (val % self.__width) + (i * self.__width)
            self._bins[t_bin] -= num_els
            if self._bins[t_bin] < self.__int32_t_min:
                self._bins[t_bin] = self.__int32_t_min
            if self._bins[t_bin] < res:
                res = self._bins[t_bin]
        self.__elements_added -= num_els
        if self.__elements_added < self.__int32_t_min:
            self.__elements_added = self.__int32_t_max

        return res

    def check(self, key, query='min'):
        ''' check number of times element 'key' is in the count-min sketch '''
        hashes = self.hashes(key)
        return self.check_alt(hashes, query)

    def check_alt(self, hashes, query='min'):
        ''' check the count-min sketch for an element by using the hashes '''
        qry = query.lower()
        bins = self.__get_values_sorted(hashes)
        if qry == 'min' or qry is 'min':
            res = bins[0]
        elif qry == 'mean' or qry is 'mean':
            res = sum(bins) // self.__depth
        elif qry == 'mean-min' or qry is 'mean-min':
            meanmin = list()
            for t_bin in bins:
                diff = self.__elements_added - t_bin
                calc = t_bin - diff // (self.__width - 1)
                meanmin.append(calc)
            meanmin.sort()
            if self.__depth % 2 == 0:
                calc = meanmin[self.__depth//2] + meanmin[self.__depth//2 - 1]
                res = calc / 2  # TODO: should this be // ?
            else:
                res = meanmin[self.__depth//2]
        else:
            print('invalid query type')
        return res

    def export(self, filepath):
        ''' export the count-min sketch to file '''
        with open(filepath, 'wb') as filepointer:
            # write out the bins
            rep = 'i' * len(self._bins)
            filepointer.write(pack(rep, *self._bins))
            filepointer.write(pack('IIq', self.__width, self.__depth,
                                   self.__elements_added))

    def load(self, filepath, hash_function=None):
        ''' load the count-min sketch from file '''
        with open(filepath, 'rb') as filepointer:
            offset = calcsize('IIq')
            filepointer.seek(offset * -1, os.SEEK_END)
            mybytes = unpack('IIq', filepointer.read(offset))
            self.__width = mybytes[0]
            self.__depth = mybytes[1]
            self.__elements_added = mybytes[2]
            self.__confidence = 1 - (1 / math.pow(2, self.__depth))
            self.__error_rate = 2 / self.__width

            filepointer.seek(0, os.SEEK_SET)
            length = self.__width * self.__depth
            rep = 'i' * length
            offset = calcsize(rep)
            self._bins = list(unpack(rep, filepointer.read(offset)))

        if hash_function is None:
            self._hash_function = self.__default_hash
        else:
            self._hash_function = hash_function

    def __default_hash(self, key, depth):
        ''' the default fnv-1a hashing routine '''
        res = list()
        tmp = key
        for _ in range(0, depth):
            if tmp != key:
                tmp = self.__fnv_1a("{0:x}".format(tmp))
            else:
                tmp = self.__fnv_1a(key)
            res.append(tmp)
        return res

    def __fnv_1a(self, key):
        ''' 64 bit fnv-1a hash '''
        hval = 14695981039346656073
        fnv_64_prime = 1099511628211
        for t_str in key:
            hval = hval ^ ord(t_str)
            hval = (hval * fnv_64_prime) % self.__uint64_t_max
        return hval

    def __get_values_sorted(self, hashes):
        ''' get the values sorted '''
        bins = list()
        for i, val in enumerate(hashes):
            t_bin = (val % self.__width) + (i * self.__width)
            bins.append(self._bins[t_bin])
        bins.sort()
        return bins


if __name__ == '__main__':
    # pass
    # Test export output
    print('build in memory check')
    cms = CountMinSketch(width=100000, depth=7)
    # add elements
    for i in range(0, 100):
        t = 100 * (i + 1)
        cms.add(str(i), t)

    print(cms.check(str(0), 'min'))
    print(cms.check(str(0), 'mean'))
    print(cms.check(str(0), 'mean-min'))
    cms.export('./dist/py_test.cms')

    print('import from disk check')
    cmsf = CountMinSketch(filepath='./dist/py_test.cms')
    if cms.width != cmsf.width:
        print('width does not match!')
    if cms.depth != cmsf.depth:
        print('depth does not match!')

    print(cmsf.check(str(0), 'min'))
    print(cmsf.check(str(0), 'mean'))
    print(cmsf.check(str(0), 'mean-min'))
